File: innovacad_api/lib/src/core/result.dart
class Result<T> {
  final T? data;
  final AppError? error;

  const Result._({this.data, this.error});

  bool get isSuccess => error == null;
  bool get isFailure => error != null;

  static Result<T> success<T>(T data) => Result._(data: data);

  static Result<T> failure<T>(AppError error) => Result._(error: error);
}

class AppError {
  final AppErrorType type;
  final String message;
  final Map<String, dynamic>? details;

  AppError(this.type, this.message, {this.details});

  Map<String, dynamic> toJson() => {
    'type': type.toString().split('.').last,
    'message': message,
    if (details != null) 'details': details,
  };
}

enum AppErrorType {
  notFound,
  validation,
  conflict,
  unauthorized,
  forbidden,
  badRequest,
  external,
  internal,
}

File: innovacad_web/src/utils/result.ts
import { AppError } from "../types/api-types";

export class Result<T> {
  public readonly data?: T;
  public readonly error?: AppError;

  private constructor(data?: T, error?: AppError) {
    this.data = data;
    this.error = error;
  }

  public get isSuccess(): boolean {
    return !this.error;
  }

  public get isFailure(): boolean {
    return !!this.error;
  }

  public static success<T>(data: T): Result<T> {
    return new Result<T>(data, undefined);
  }
  public static failure<T>(error: AppError): Result<T> {
    return new Result<T>(undefined, error);
  }
}
